### PR TITLE
default session implements persistable interface, can save SQL variables

### PR DESCRIPTION
**Pull request by max-hoffman**

### CODE REVIEW

File: go/libraries/doltcore/sqle/dsess/session.go [hunk 1]
```diff
@@ -232,16 +231,34 @@ func NewSession(ctx *sql.Context, sqlSess sql.Session, pro RevisionDatabaseProvi
 	return sess, nil
 }
 
+// NewDoltSession creates a DoltSession object from a standard sql.Session and 0 or more Database objects.
+func NewDoltSession(ctx *sql.Context, sqlSess *sql.BaseSession, pro RevisionDatabaseProvider, conf *env.DoltCliConfig, dbs ...InitialDbState) (*DoltSession, error) {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dsess/session.go:235]: Why is this here instead of in doltsession.go?

File: go/libraries/doltcore/sqle/sqlpersist_test.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/store/types"
+)
+
+// Structure for a test of a insert query
+type PersistTest struct {
+	// The name of this test. Names should be unique and descriptive.
+	Name string
+	// The insert query to run
+	PersitQuery string
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/sqlpersist_test.go:36]: typo

File: go/libraries/utils/config/prefix_config.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+package config
+
+import (
+	"fmt"
+	"strings"
+)
+
+// PrefixConfig decorates read and write access to the underlying config by appending a prefix to the accessed keys
+// on reads and writes
+// TODO: this is temporary to namespace server configs in .dolt/config.json
```
Comment (zachmu) [inline go/libraries/utils/config/prefix_config.go:24]: TODOs still relevant?

File: go/libraries/doltcore/sqle/dsess/dolt_session.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+type DoltSession struct {
+	*Session
+	globalsConf config.ReadWriteConfig
+	mu          *sync.Mutex
+}
+
+var _ sql.Session = (*DoltSession)(nil)
+var _ sql.PersistableSession = (*DoltSession)(nil)
+
+func NewDoltSessionFromDefault(sess *Session, conf config.ReadWriteConfig) *DoltSession {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dsess/dolt_session.go:38]: FromDefault doesn't really make sense here

WithConfig?

File: go/libraries/doltcore/sqle/dsess/dolt_session.go [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+		return config.SetFloat(conf, key, v)
+	case string:
+		return config.SetString(conf, key, v)
+	default:
+		return sql.ErrInvalidType.New(v)
+	}
+}
+
+// NewPersistedSystemVariables returns system variables from the persisted config
+func NewPersistedSystemVariables(conf config.ReadableConfig) ([]sql.SystemVariable, error) {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dsess/dolt_session.go:171]: Call this SystemVariablesInConfig

File: go/libraries/doltcore/sqle/dsess/dolt_session.go [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+		i++
+		return false
+	})
+	if err != nil {
+		return nil, err
+	}
+	return allVars, nil
+}
+
+func InitPersistedSystemVars(dEnv *env.DoltEnv) error {
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/dsess/dolt_session.go:196]: This method makes me uneasy. Seems like calling this more than once, or after an engine is already server queries, is really bad news?

Needs a lock to make sure it's only called once

**Review by zachmu — APPROVED** [review]
Looks pretty good, just a few comments

File: go/cmd/dolt/commands/sqlserver/server.go [hunk 7]
```diff
@@ -83,6 +84,26 @@ func Serve(ctx context.Context, version string, serverConfig ServerConfig, serve
 		permissions = auth.ReadPerm
 	}
 
+	// Do not set the value of Version.  Let it default to what go-mysql-server uses.  This should be equivalent
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/cmd/dolt/commands/sqlserver/server.go:86]: was in constructor previously. i'll move it to where most of the vals are set

File: go/cmd/dolt/commands/config_test.go [hunk 8]
```diff
@@ -19,14 +19,260 @@ import (
...
+		globalCfg.SetStrings(map[string]string{"title": "senior dufus"})
+	case env.LocalConfig:
+		dEnv.Config.CreateLocalConfig(map[string]string{"title": "senior dufus"})
+	}
+}
+func TestConfigAdd(t *testing.T) {
+	tests := []struct {
+		Name       string
+		CfgSet     *set.StrSet
+		ConfigElem env.DoltConfigElement
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/cmd/dolt/commands/config_test.go:46]: env.ConfigScope?

File: go/libraries/doltcore/sqle/dsess/session.go [hunk 9]
```diff
@@ -232,16 +231,34 @@ func NewSession(ctx *sql.Context, sqlSess sql.Session, pro RevisionDatabaseProvi
...
+	sess, err := NewSession(ctx, sqlSess, pro, conf, dbs...)
+	if err != nil {
+		return nil, err
+	}
+
+	var globals config.ReadWriteConfig
+	if localConf, ok := conf.GetConfig(env.LocalConfig); !ok {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dsess/session.go:242]: i'll try to figure out something more direct, 90% sure there is an edge case for running multi-db-mode from a valid dolt dir that can create unpredictable behavior, but the multidb code wasn't super organized to begin with so it felt like premature optimization

File: go/libraries/doltcore/sqle/dsess/dolt_session.go (hunk 4 above)
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dsess/dolt_session.go:38]: this inputs a`NewDefaultSession`, but i see how that can be confusing. How about `dsess.Session.NewDoltSession(conf)`?

File: go/cmd/dolt/commands/sqlserver/serverconfig.go [hunk 10]
```diff
@@ -101,6 +102,8 @@ type ServerConfig interface {
 	TLSCert() string
 	// RequireSecureTransport is true if the server should reject non-TLS connections.
 	RequireSecureTransport() bool
+	// NoDefaults is true if we do not load persisted globals
+	NoDefaults() bool
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/cmd/dolt/commands/sqlserver/serverconfig.go:106]: now `--persistence-behavior={<empty>, load, ignore}`

File: go/libraries/doltcore/sqle/dsess/dolt_session.go (hunk 6 above)
Comment (max-hoffman) [reply go/libraries/doltcore/sqle/dsess/dolt_session.go:196]: yeah there were some ugly race's on the first pass. adding a global `initMu` lock, lmk if that's too much of an antipattern
